tauri-app/src-tauri/src/commands/fork.rs

```diff
@@ -0,0 +1,279 @@
+use alloy_dyn_abi::JsonAbiExt;
+use alloy_json_abi::Error;
+use forker::*;
+use once_cell::sync::Lazy;
+use reqwest::Client;
+use revm::primitives::Bytes;
+use serde_bytes::ByteBuf;
+use std::{collections::HashMap, sync::Mutex};
+
+const SELECTOR_REGISTRY_URL: &str = "https://api.openchain.xyz/signature-database/v1/lookup";
+
+/// static hashmap of fork evm instances, used for caching instances between runs
+pub static FORKS: Lazy<Mutex<HashMap<String, ForkedEvm>>> =
+    Lazy::new(|| Mutex::new(HashMap::new()));
+
+/// hashmap of cached error selectors    
+pub static SELECTORS: Lazy<Mutex<HashMap<[u8; 4], Error>>> =
+    Lazy::new(|| Mutex::new(HashMap::new()));
+
+#[derive(Debug, Clone, PartialEq, serde::Serialize, serde::Deserialize)]
+pub enum DecodedErrorType {
+    Unknown,
+    Known {
+        name: String,
+        args: Vec<String>,
+        sig: String,
+    },
+}
+
+#[tauri::command]
+pub async fn fork_call(
+    fork_url: &str,
+    fork_block_number: u64,
+    from_address: ByteBuf,
+    to_address: ByteBuf,
+    calldata: ByteBuf,
+) -> Result<Result<Bytes, DecodedErrorType>, String> {
+    // build key from fork url and block number
+    let key = fork_url.to_owned() + &fork_block_number.to_string();
+
+    let is_cached = { FORKS.lock().unwrap().contains_key(&key) };
+
+    let result = if is_cached {
+        let mut forks = FORKS.lock().unwrap();
+        let forked_evm = forks.get_mut(&key).unwrap();
+
+        // call a contract read-only
+        forked_evm
+            .call(
+                from_address.as_slice(),
+                to_address.as_slice(),
+                calldata.as_slice(),
+            )
+            .map_err(|e| e.to_string())?
+    } else {
+        let mut forked_evm =
+            ForkedEvm::new(fork_url, Some(fork_block_number), Some(200000u64), None).await;
+
+        // call a contract read-only
+        let res = forked_evm
+            .call(
+                from_address.as_slice(),
+                to_address.as_slice(),
+                calldata.as_slice(),
+            )
+            .map_err(|e| e.to_string())?;
+
+        // lock static FORKS
+        let mut forks = FORKS.lock().unwrap();
+        forks.insert(key, forked_evm);
+        res
+    };
+
+    if result.reverted {
+        // decode result bytes to error selectors if it was a revert
+        Ok(Err(decode_error(&result.result).await?))
+    } else {
+        Ok(Ok(result.result))
+    }
+}
+
+/// decodes an error returned from calling a contract by searching its selector in registry
+async fn decode_error(error_data: &[u8]) -> Result<DecodedErrorType, String> {
```

Comment:
mind if we move this function into the `common` crate? I'd like to use it in the cli too.

Reply:
or maybe it belongs in the WritableClient in alloy-ethers-typecast -- we can save that for a separate PR though

Reply:
yes sure